File: app/src/main/java/com/walkercoding/videocompress/videocompress/FrameCompressor.java
package com.walkercoding.videocompress.videocompress;

import android.media.MediaCodec;
import android.media.MediaCodecInfo;
import android.media.MediaFormat;
import android.os.Build;
import com.walkercoding.videocompress.videocompress.helpers.InputSurface;
import com.walkercoding.videocompress.videocompress.helpers.OutputSurface;
import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

import java.io.IOException;
import java.nio.ByteBuffer;

import static com.walkercoding.videocompress.videocompress.CompressInfo.MIME_TYPE;
import static com.walkercoding.videocompress.videocompress.CompressInfo.TIMEOUT_USEC;

public class FrameCompressor {

    private final static int PROCESSOR_TYPE_OTHER = 0;
    private final static int PROCESSOR_TYPE_QCOM = 1;
    private final static int PROCESSOR_TYPE_INTEL = 2;
    private final static int PROCESSOR_TYPE_MTK = 3;
    private final static int PROCESSOR_TYPE_SEC = 4;
    private final static int PROCESSOR_TYPE_TI = 5;

    private MediaCodec decoder = null;
    private MediaCodec encoder = null;
    private InputSurface inputSurface = null;
    private OutputSurface outputSurface = null;

    private boolean outputDone = false;
    private boolean inputDone = false;
    private boolean decoderDone = false;

    private static final int INVALID_VIDEO_TRACK_INDEX = -5;
    private int videoTrackIndex = INVALID_VIDEO_TRACK_INDEX;

    private ByteBuffer[] decoderInputBuffers = null;
    private ByteBuffer[] encoderOutputBuffers = null;
    private ByteBuffer[] encoderInputBuffers = null;

    private MediaCodec.BufferInfo info = new MediaCodec.BufferInfo();
    private long videoStartTime = -1;

    private void initCoders(ContainerConverter.Param containerParam, CompressInfo compressInfo, DeviceOption deviceOption) throws IOException {
        // setup encoder
        encoder = MediaCodec.createEncoderByType(MIME_TYPE);
        encoder.configure(containerParam.getOutputFormat(compressInfo, deviceOption.colorFormat), null, null, MediaCodec.CONFIGURE_FLAG_ENCODE);
        if (Build.VERSION.SDK_INT >= 18) {
            inputSurface = new InputSurface(encoder.createInputSurface());
            inputSurface.makeCurrent();
        }
        encoder.start();
        // setup decoder
        decoder = MediaCodec.createDecoderByType(containerParam.getInputFormat().getString(MediaFormat.KEY_MIME));
        if (Build.VERSION.SDK_INT >= 18) {
            outputSurface = new OutputSurface();
        } else {
            outputSurface = new OutputSurface(compressInfo.resultWidth, compressInfo.resultHeight, compressInfo.rotateRender);
        }
        decoder.configure(containerParam.getInputFormat(), outputSurface.getSurface(), null, 0);
        decoder.start();
    }

    private ByteBuffer getDecoderInputBuffer(int index) {
        if (Build.VERSION.SDK_INT < 21) {
            if (decoderInputBuffers == null) {
                decoderInputBuffers = decoder.getInputBuffers();
            }
            return decoderInputBuffers[index];
        } else {
            return decoder.getInputBuffer(index);
        }
    }

    private ByteBuffer getEncodeOutputBuffer(int index) {
        if (Build.VERSION.SDK_INT < 21) {
            if (encoderOutputBuffers == null) {
                encoderOutputBuffers = encoder.getOutputBuffers();
            }
            return encoderOutputBuffers[index];
        } else {
            return encoder.getOutputBuffer(index);
        }
    }

    private ByteBuffer getEncoderInputBuffers(int index) {
        if (encoderInputBuffers == null) {
            encoderInputBuffers = encoder.getInputBuffers();
        }
        return encoderInputBuffers[index];
    }

    private void fromExtractorToDecoder(ContainerConverter.Param containerParam) {
        if (!inputDone) {
            final int index = containerParam.getSampleTrackIndex();
            if (index == containerParam.videoIndex) {
                int inputBufIndex = decoder.dequeueInputBuffer(TIMEOUT_USEC);
                if (inputBufIndex >= 0) {
                    ByteBuffer inputBuf = getDecoderInputBuffer(inputBufIndex);
                    int chunkSize = containerParam.readSampleData(inputBuf);
                    if (chunkSize < 0) {
                        finishDecoder(inputBufIndex);
                    } else {
                        decoder.queueInputBuffer(inputBufIndex, 0, chunkSize, containerParam.getSampleTime(), 0);
                        containerParam.advance();
                    }
                }
            } else if (index == -1) {
                // EOF
                int inputBufIndex = decoder.dequeueInputBuffer(TIMEOUT_USEC);
                if (inputBufIndex >= 0) {
                    finishDecoder(inputBufIndex);
                }
            }
        }
    }

    private void finishDecoder(int inputBufIndex) {
        decoder.queueInputBuffer(inputBufIndex, 0, 0, 0L, MediaCodec.BUFFER_FLAG_END_OF_STREAM);
        inputDone = true;
    }

    /**
     * @return true -- encoder doesn't has pending surface that need be written to muxer
     * so need feed new surface to encoder
     */
    private boolean fromEncoderToMuxer(ContainerConverter.Param containerParam, CompressInfo compressInfo) throws Exception {
        int encoderOutputIndex = encoder.dequeueOutputBuffer(info, TIMEOUT_USEC);
        if (encoderOutputIndex == MediaCodec.INFO_TRY_AGAIN_LATER) {
            return true;
        }
        if (encoderOutputIndex == MediaCodec.INFO_OUTPUT_BUFFERS_CHANGED) {
            if (Build.VERSION.SDK_INT < 21) {
                // invalid encoderOutputBuffers
                encoderOutputBuffers = null;
            }
            return false;
        }
        if (encoderOutputIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
            MediaFormat newFormat = encoder.getOutputFormat();
            if (videoTrackIndex == INVALID_VIDEO_TRACK_INDEX) {
                videoTrackIndex = containerParam.addTrack(newFormat);
            }
            return false;
        }
        if (encoderOutputIndex < 0) {
            throw new RuntimeException("unexpected result from encoder.dequeueOutputBuffer: " + encoderOutputIndex);
        }
        ByteBuffer encodedData = getEncodeOutputBuffer(encoderOutputIndex);
        if (encodedData == null) {
            throw new RuntimeException("encoderOutputBuffer " + encoderOutputIndex + " was null");
        }
        if (info.size > 1) {
            if ((info.flags & MediaCodec.BUFFER_FLAG_CODEC_CONFIG) == 0) {
                if (containerParam.writeSampleData(videoTrackIndex, encodedData, info, true)) {
                    compressInfo.didWriteData(false, false);
                }
            } else if (videoTrackIndex == INVALID_VIDEO_TRACK_INDEX) {
                addVideoTrackToMuxer(compressInfo, containerParam, encodedData);
            }
        }
        outputDone = (info.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0;
        encoder.releaseOutputBuffer(encoderOutputIndex, false);
        return false;
    }

    private void addVideoTrackToMuxer(CompressInfo compressInfo, ContainerConverter.Param containerParam, ByteBuffer encodedData) {
        byte[] csd = new byte[info.size];
        encodedData.limit(info.offset + info.size);
        encodedData.position(info.offset);
        encodedData.get(csd);
        ByteBuffer sps = null;
        ByteBuffer pps = null;
        for (int a = info.size - 1; a >= 0; a--) {
            if (a > 3) {
                if (csd[a] == 1 && csd[a - 1] == 0 && csd[a - 2] == 0 && csd[a - 3] == 0) {
                    sps = ByteBuffer.allocate(a - 3);
                    pps = ByteBuffer.allocate(info.size - (a - 3));
                    sps.put(csd, 0, a - 3).position(0);
                    pps.put(csd, a - 3, info.size - (a - 3)).position(0);
                    break;
                }
            } else {
                break;
            }
        }
        MediaFormat newFormat = MediaFormat.createVideoFormat(MIME_TYPE, compressInfo.resultWidth, compressInfo.resultHeight);
        if (sps != null) {
            newFormat.setByteBuffer("csd-0", sps);
            newFormat.setByteBuffer("csd-1", pps);
        }
        videoTrackIndex = containerParam.addTrack(newFormat);
    }

    private void fromDecoderToEncoder(CompressInfo compressInfo, DeviceOption deviceOption, AdjustSize adjustSize) {
        if (!decoderDone) {
            int decoderStatus = decoder.dequeueOutputBuffer(info, TIMEOUT_USEC);
            if (decoderStatus == MediaCodec.INFO_TRY_AGAIN_LATER
                    || decoderStatus == MediaCodec.INFO_OUTPUT_BUFFERS_CHANGED) {
                return;
            }
            if (decoderStatus == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
                MediaFormat newFormat = decoder.getOutputFormat();
                LogUtils.e("newFormat = " + newFormat);
                return;
            }
            if (decoderStatus < 0) {
                throw new RuntimeException("unexpected result from decoder.dequeueOutputBuffer: " + decoderStatus);
            }
            boolean doRender = info.size != 0 || (Build.VERSION.SDK_INT < 18 && info.presentationTimeUs != 0);
            if (compressInfo.endTime > 0 && info.presentationTimeUs >= compressInfo.endTime) {
                // cut tail
                inputDone = true;
                decoderDone = true;
                doRender = false;
                info.flags |= MediaCodec.BUFFER_FLAG_END_OF_STREAM;
            }
            if (compressInfo.startTime > 0 && videoStartTime == -1) {
                // cut header
                if (info.presentationTimeUs < compressInfo.startTime) {
                    doRender = false;
                    LogUtils.e("drop frame startTime = " + compressInfo.startTime + " present time = " + info.presentationTimeUs);
                } else {
                    videoStartTime = info.presentationTimeUs;
                }
            }
            decoder.releaseOutputBuffer(decoderStatus, doRender);
            if (doRender) {
                try {
                    outputSurface.awaitNewImage();
                    if (Build.VERSION.SDK_INT >= 18) {
                        outputSurface.drawImage(false);
                        inputSurface.setPresentationTime(info.presentationTimeUs * 1000);
                        inputSurface.swapBuffers();
                    } else {
                        int inputBufIndex = encoder.dequeueInputBuffer(TIMEOUT_USEC);
                        if (inputBufIndex >= 0) {
                            outputSurface.drawImage(true);
                            ByteBuffer rgbBuf = outputSurface.getFrame();
                            ByteBuffer yuvBuf = getEncoderInputBuffers(inputBufIndex);
                            yuvBuf.clear();
                            JniUtils.convertVideoFrame(rgbBuf, yuvBuf, deviceOption.colorFormat, compressInfo.resultWidth, compressInfo.resultHeight, adjustSize.padding, deviceOption.swapUV);
                            encoder.queueInputBuffer(inputBufIndex, 0, adjustSize.bufferSize, info.presentationTimeUs, 0);
                        } else {
                            LogUtils.e("input buffer not available");
                        }
                    }
                } catch (Exception e) {
                    LogUtils.e(e);
                }
            }
            if ((info.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) {
                LogUtils.e("decoder stream end");
                if (Build.VERSION.SDK_INT >= 18) {
                    encoder.signalEndOfInputStream();
                } else {
                    int inputBufIndex = encoder.dequeueInputBuffer(TIMEOUT_USEC);
                    if (inputBufIndex >= 0) {
                        encoder.queueInputBuffer(inputBufIndex, 0, 1, info.presentationTimeUs, MediaCodec.BUFFER_FLAG_END_OF_STREAM);
                    }
                }
            }
        }
    }

    long run(final CompressInfo compressInfo, final ContainerConverter.Param containerParam) {
        try {
            final DeviceOption deviceOption = getDeviceOption();
            final AdjustSize adjustSize = getAdjustSize(compressInfo, deviceOption);

            containerParam.selectVideoIndex();
            containerParam.seekIfNeed(compressInfo.startTime);
            initCoders(containerParam, compressInfo, deviceOption);

            final Consumer<Throwable> onError = new Consumer<Throwable>() {
                @Override
                public void accept(Throwable throwable) throws Exception {
                    if (throwable instanceof Exception) {
                        throw (Exception) throwable;
                    }
                }
            };
            Observable.create(new ObservableOnSubscribe<Object>() {
                @Override
                public void subscribe(ObservableEmitter<Object> emitter) throws Exception {
                    while (!inputDone) {
                        fromExtractorToDecoder(containerParam);
                    }
                }
            })
                    .subscribeOn(Schedulers.newThread())
                    .doOnError(onError)
                    .subscribe();

            while (!outputDone) {
                // write encoder data to muxer first, then read to encoder from decoder
                if (fromEncoderToMuxer(containerParam, compressInfo)) {
                    if (!decoderDone) {
                        fromDecoderToEncoder(compressInfo, deviceOption, adjustSize);
                    }
                }
            }
        } catch (Exception e) {
            LogUtils.e(e);
            compressInfo.error = true;
        }

        containerParam.unselectVideoIndex();

        if (outputSurface != null) {
            outputSurface.release();
        }
        if (inputSurface != null) {
            inputSurface.release();
        }
        if (decoder != null) {
            decoder.stop();
            decoder.release();
        }
        if (encoder != null) {
            encoder.stop();
            encoder.release();
        }

        return videoStartTime;
    }

    private static class DeviceOption {
        private int colorFormat;
        private int processorType = PROCESSOR_TYPE_OTHER;
        private String manufacturer = Build.MANUFACTURER.toLowerCase();
        private int swapUV = 0;
    }

    private DeviceOption getDeviceOption() {
        DeviceOption option = new DeviceOption();
        if (Build.VERSION.SDK_INT < 18) {
            MediaCodecInfo codecInfo = Utils.selectCodec(MIME_TYPE);
            option.colorFormat = Utils.selectColorFormat(codecInfo, MIME_TYPE);
            if (option.colorFormat == 0) {
                throw new RuntimeException("no supported color format");
            }
            String codecName = codecInfo.getName();
            if (codecName.contains("OMX.qcom.")) {
                option.processorType = PROCESSOR_TYPE_QCOM;
                if (Build.VERSION.SDK_INT == 16) {
                    if (option.manufacturer.equals("lge") || option.manufacturer.equals("nokia")) {
                        option.swapUV = 1;
                    }
                }
            } else if (codecName.contains("OMX.Intel.")) {
                option.processorType = PROCESSOR_TYPE_INTEL;
            } else if (codecName.equals("OMX.MTK.VIDEO.ENCODER.AVC")) {
                option.processorType = PROCESSOR_TYPE_MTK;
            } else if (codecName.equals("OMX.SEC.AVC.Encoder")) {
                option.processorType = PROCESSOR_TYPE_SEC;
                option.swapUV = 1;
            } else if (codecName.equals("OMX.TI.DUCATI1.VIDEO.H264E")) {
                option.processorType = PROCESSOR_TYPE_TI;
            }
            LogUtils.e("codec = " + codecInfo.getName() + " manufacturer = " + option.manufacturer + "device = " + Build.MODEL);
        } else {
            option.colorFormat = MediaCodecInfo.CodecCapabilities.COLOR_FormatSurface;
        }
        LogUtils.e("colorFormat = " + option.colorFormat);
        return option;
    }

    private static class AdjustSize {
        private int padding = 0;
        private int bufferSize;
    }

    private AdjustSize getAdjustSize(CompressInfo compressInfo, DeviceOption deviceOption) {
        AdjustSize adjustSize = new AdjustSize();
        int resultHeightAligned = compressInfo.resultHeight;
        adjustSize.bufferSize = compressInfo.resultWidth * compressInfo.resultHeight * 3 / 2;

        if (deviceOption.processorType == PROCESSOR_TYPE_OTHER) {
            if (compressInfo.resultHeight % 16 != 0) {
                resultHeightAligned += (16 - (compressInfo.resultHeight % 16));
                adjustSize.padding = compressInfo.resultWidth * (resultHeightAligned - compressInfo.resultHeight);
                adjustSize.bufferSize += adjustSize.padding * 5 / 4;
            }
        } else if (deviceOption.processorType == PROCESSOR_TYPE_QCOM) {
            if (!deviceOption.manufacturer.toLowerCase().equals("lge")) {
                int uvoffset = (compressInfo.resultWidth * compressInfo.resultHeight + 2047) & ~2047;
                adjustSize.padding = uvoffset - (compressInfo.resultWidth * compressInfo.resultHeight);
                adjustSize.bufferSize += adjustSize.padding;
            }
        } else if (deviceOption.processorType == PROCESSOR_TYPE_TI) {
            //resultHeightAligned = 368;
            //bufferSize = resultWidth * resultHeightAligned * 3 / 2;
            //resultHeightAligned += (16 - (resultHeight % 16));
            //padding = resultWidth * (resultHeightAligned - resultHeight);
            //bufferSize += padding * 5 / 4;
        } else if (deviceOption.processorType == PROCESSOR_TYPE_MTK) {
            if (deviceOption.manufacturer.equals("baidu")) {
                resultHeightAligned += (16 - (compressInfo.resultHeight % 16));
                adjustSize.padding = compressInfo.resultWidth * (resultHeightAligned - compressInfo.resultHeight);
                adjustSize.bufferSize += adjustSize.padding * 5 / 4;
            }
        }

        return adjustSize;
    }
}

File: app/src/main/java/com/walkercoding/videocompress/videocompress/FrameCompressor.kt
package com.walkercoding.videocompress.videocompress

import android.media.MediaCodec
import android.media.MediaCodecInfo
import android.media.MediaFormat
import android.os.Build
import com.walkercoding.videocompress.videocompress.CompressInfo.Companion.MIME_TYPE
import com.walkercoding.videocompress.videocompress.CompressInfo.Companion.TIMEOUT_USEC
import com.walkercoding.videocompress.videocompress.helpers.InputSurface
import com.walkercoding.videocompress.videocompress.helpers.OutputSurface
import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe
import io.reactivex.schedulers.Schedulers
import java.io.IOException
import java.nio.ByteBuffer

class FrameCompressor {

    private var decoder: MediaCodec? = null
    private var encoder: MediaCodec? = null
    private var inputSurface: InputSurface? = null
    private var outputSurface: OutputSurface? = null

    private var outputDone = false
    private var inputDone = false
    private var decoderDone = false
    private var videoTrackIndex = INVALID_VIDEO_TRACK_INDEX

    private var decoderInputBuffers: Array<ByteBuffer>? = null
    private var encoderOutputBuffers: Array<ByteBuffer>? = null
    private var encoderInputBuffers: Array<ByteBuffer>? = null

    private val info = MediaCodec.BufferInfo()
    private var videoStartTime: Long = -1

    private val deviceOption: DeviceOption
        get() {
            val option = DeviceOption()
            if (Build.VERSION.SDK_INT < 18) {
                val codecInfo = Utils.selectCodec(MIME_TYPE)
                option.colorFormat = Utils.selectColorFormat(codecInfo!!, MIME_TYPE)
                if (option.colorFormat == 0) {
                    throw RuntimeException("no supported color format")
                }
                val codecName = codecInfo.name
                if (codecName.contains("OMX.qcom.")) {
                    option.processorType = PROCESSOR_TYPE_QCOM
                    if (Build.VERSION.SDK_INT == 16) {
                        if (option.manufacturer == "lge" || option.manufacturer == "nokia") {
                            option.swapUV = 1
                        }
                    }
                } else if (codecName.contains("OMX.Intel.")) {
                    option.processorType = PROCESSOR_TYPE_INTEL
                } else if (codecName == "OMX.MTK.VIDEO.ENCODER.AVC") {
                    option.processorType = PROCESSOR_TYPE_MTK
                } else if (codecName == "OMX.SEC.AVC.Encoder") {
                    option.processorType = PROCESSOR_TYPE_SEC
                    option.swapUV = 1
                } else if (codecName == "OMX.TI.DUCATI1.VIDEO.H264E") {
                    option.processorType = PROCESSOR_TYPE_TI
                }
                LogUtils.i("codec = " + codecInfo.name + " manufacturer = " + option.manufacturer + "device = " + Build.MODEL)
            } else {
                option.colorFormat = MediaCodecInfo.CodecCapabilities.COLOR_FormatSurface
            }
            LogUtils.i("colorFormat = " + option.colorFormat)
            return option
        }

    @Throws(IOException::class)
    private fun initCoders(containerParam: ContainerConverter.Param, compressInfo: CompressInfo, deviceOption: DeviceOption) {
        // setup encoder
        encoder = MediaCodec.createEncoderByType(MIME_TYPE)
        encoder!!.configure(containerParam.getOutputFormat(compressInfo, deviceOption.colorFormat), null, null, MediaCodec.CONFIGURE_FLAG_ENCODE)
        if (Build.VERSION.SDK_INT >= 18) {
            inputSurface = InputSurface(encoder!!.createInputSurface())
            inputSurface!!.makeCurrent()
        }
        encoder!!.start()
        // setup decoder
        decoder = MediaCodec.createDecoderByType(containerParam.getInputFormat().getString(MediaFormat.KEY_MIME))
        if (Build.VERSION.SDK_INT >= 18) {
            outputSurface = OutputSurface()
        } else {
            outputSurface = OutputSurface(compressInfo.resultWidth, compressInfo.resultHeight, compressInfo.rotateRender)
        }
        decoder!!.configure(containerParam.getInputFormat(), outputSurface!!.surface, null, 0)
        decoder!!.start()
    }

    private fun getDecoderInputBuffer(index: Int): ByteBuffer {
        if (Build.VERSION.SDK_INT < 21) {
            if (decoderInputBuffers == null) {
                decoderInputBuffers = decoder!!.inputBuffers
            }
            return decoderInputBuffers!![index]
        } else {
            return decoder!!.getInputBuffer(index)
        }
    }

    private fun getEncodeOutputBuffer(index: Int): ByteBuffer? {
        if (Build.VERSION.SDK_INT < 21) {
            if (encoderOutputBuffers == null) {
                encoderOutputBuffers = encoder!!.outputBuffers
            }
            return encoderOutputBuffers!![index]
        } else {
            return encoder!!.getOutputBuffer(index)
        }
    }

    private fun getEncoderInputBuffers(index: Int): ByteBuffer {
        if (encoderInputBuffers == null) {
            encoderInputBuffers = encoder!!.inputBuffers
        }
        return encoderInputBuffers!![index]
    }

    private fun decodeFromExtractor(containerParam: ContainerConverter.Param) {
        if (!inputDone) {
            val index = containerParam.sampleTrackIndex
            if (index == containerParam.videoIndex) {
                val inputBufIndex = decoder!!.dequeueInputBuffer(TIMEOUT_USEC.toLong())
                if (inputBufIndex >= 0) {
                    val inputBuf = getDecoderInputBuffer(inputBufIndex)
                    val chunkSize = containerParam.readSampleData(inputBuf)
                    if (chunkSize < 0) {
                        finishDecoder(inputBufIndex)
                    } else {
                        decoder!!.queueInputBuffer(inputBufIndex, 0, chunkSize, containerParam.sampleTime, 0)
                        containerParam.advance()
                    }
                }
            } else if (index == -1) {
                // EOF
                val inputBufIndex = decoder!!.dequeueInputBuffer(TIMEOUT_USEC.toLong())
                if (inputBufIndex >= 0) {
                    finishDecoder(inputBufIndex)
                }
            }
        }
    }

    private fun finishDecoder(inputBufIndex: Int) {
        decoder!!.queueInputBuffer(inputBufIndex, 0, 0, 0L, MediaCodec.BUFFER_FLAG_END_OF_STREAM)
        inputDone = true
    }

    /**
     * @return true -- encoder doesn't has pending surface that need be written to muxer
     * so need feed new surface to encoder
     */
    @Throws(Exception::class)
    private fun encodeToMuxer(containerParam: ContainerConverter.Param, compressInfo: CompressInfo): Boolean {
        val encoderOutputIndex = encoder!!.dequeueOutputBuffer(info, TIMEOUT_USEC.toLong())
        if (encoderOutputIndex == MediaCodec.INFO_TRY_AGAIN_LATER) {
            return true
        }
        if (encoderOutputIndex == MediaCodec.INFO_OUTPUT_BUFFERS_CHANGED) {
            if (Build.VERSION.SDK_INT < 21) {
                // invalid encoderOutputBuffers
                encoderOutputBuffers = null
            }
            return false
        }
        if (encoderOutputIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
            val newFormat = encoder!!.outputFormat
            if (videoTrackIndex == INVALID_VIDEO_TRACK_INDEX) {
                videoTrackIndex = containerParam.addTrack(newFormat)
            }
            return false
        }
        if (encoderOutputIndex < 0) {
            throw RuntimeException("unexpected result from encoder.dequeueOutputBuffer: $encoderOutputIndex")
        }
        val encodedData = getEncodeOutputBuffer(encoderOutputIndex)
                ?: throw RuntimeException("encoderOutputBuffer $encoderOutputIndex was null")
        if (info.size > 1) {
            if (info.flags and MediaCodec.BUFFER_FLAG_CODEC_CONFIG == 0) {
                if (containerParam.writeSampleData(videoTrackIndex, encodedData, info, true)) {
                    compressInfo.listener.onProgress(info.presentationTimeUs / (compressInfo.durationUs.toFloat()))
                }
            } else if (videoTrackIndex == INVALID_VIDEO_TRACK_INDEX) {
                addVideoTrackToMuxer(compressInfo, containerParam, encodedData)
            }
        }
        outputDone = info.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM != 0
        encoder!!.releaseOutputBuffer(encoderOutputIndex, false)
        return false
    }

    private fun addVideoTrackToMuxer(compressInfo: CompressInfo, containerParam: ContainerConverter.Param, encodedData: ByteBuffer) {
        val csd = ByteArray(info.size)
        encodedData.limit(info.offset + info.size)
        encodedData.position(info.offset)
        encodedData.get(csd)
        var sps: ByteBuffer? = null
        var pps: ByteBuffer? = null
        for (a in info.size - 1 downTo 0) {
            if (a > 3) {
                if (csd[a].toInt() == 1 && csd[a - 1].toInt() == 0 && csd[a - 2].toInt() == 0 && csd[a - 3].toInt() == 0) {
                    sps = ByteBuffer.allocate(a - 3)
                    pps = ByteBuffer.allocate(info.size - (a - 3))
                    sps!!.put(csd, 0, a - 3).position(0)
                    pps!!.put(csd, a - 3, info.size - (a - 3)).position(0)
                    break
                }
            } else {
                break
            }
        }
        val newFormat = MediaFormat.createVideoFormat(MIME_TYPE, compressInfo.resultWidth, compressInfo.resultHeight)
        if (sps != null) {
            newFormat.setByteBuffer("csd-0", sps)
            newFormat.setByteBuffer("csd-1", pps)
        }
        videoTrackIndex = containerParam.addTrack(newFormat)
    }

    private fun decoderToEncoder(compressInfo: CompressInfo, deviceOption: DeviceOption, adjustSize: AdjustSize) {
        if (!decoderDone) {
            val decoderStatus = decoder!!.dequeueOutputBuffer(info, TIMEOUT_USEC.toLong())
            if (decoderStatus == MediaCodec.INFO_TRY_AGAIN_LATER || decoderStatus == MediaCodec.INFO_OUTPUT_BUFFERS_CHANGED) {
                return
            }
            if (decoderStatus == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
                val newFormat = decoder!!.outputFormat
                LogUtils.i("newFormat = $newFormat")
                return
            }
            if (decoderStatus < 0) {
                throw RuntimeException("unexpected result from decoder.dequeueOutputBuffer: $decoderStatus")
            }
            var doRender = info.size != 0 || Build.VERSION.SDK_INT < 18 && info.presentationTimeUs != 0L
            if (compressInfo.endTime > 0 && info.presentationTimeUs >= compressInfo.endTime) {
                // cut tail
                inputDone = true
                decoderDone = true
                doRender = false
                info.flags = info.flags or MediaCodec.BUFFER_FLAG_END_OF_STREAM
            }
            if (compressInfo.startTime > 0 && videoStartTime == -1L) {
                // cut header
                if (info.presentationTimeUs < compressInfo.startTime) {
                    doRender = false
                    LogUtils.i("drop frame startTime = " + compressInfo.startTime + " present time = " + info.presentationTimeUs)
                } else {
                    videoStartTime = info.presentationTimeUs
                }
            }
            decoder!!.releaseOutputBuffer(decoderStatus, doRender)
            if (doRender) {
                try {
                    outputSurface!!.awaitNewImage()
                    if (Build.VERSION.SDK_INT >= 18) {
                        outputSurface!!.drawImage(false)
                        inputSurface!!.setPresentationTime(info.presentationTimeUs * 1000)
                        inputSurface!!.swapBuffers()
                    } else {
                        val inputBufIndex = encoder!!.dequeueInputBuffer(TIMEOUT_USEC.toLong())
                        if (inputBufIndex >= 0) {
                            outputSurface!!.drawImage(true)
                            val rgbBuf = outputSurface!!.frame
                            val yuvBuf = getEncoderInputBuffers(inputBufIndex)
                            yuvBuf.clear()
                            JniUtils.convertVideoFrame(rgbBuf, yuvBuf, deviceOption.colorFormat, compressInfo.resultWidth, compressInfo.resultHeight, adjustSize.padding, deviceOption.swapUV)
                            encoder!!.queueInputBuffer(inputBufIndex, 0, adjustSize.bufferSize, info.presentationTimeUs, 0)
                        } else {
                            LogUtils.i("input buffer not available")
                        }
                    }
                } catch (e: Exception) {
                    LogUtils.e(e)
                }

            }
            if (info.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM != 0) {
                LogUtils.i("decoder stream end")
                if (Build.VERSION.SDK_INT >= 18) {
                    encoder!!.signalEndOfInputStream()
                } else {
                    val inputBufIndex = encoder!!.dequeueInputBuffer(TIMEOUT_USEC.toLong())
                    if (inputBufIndex >= 0) {
                        encoder!!.queueInputBuffer(inputBufIndex, 0, 1, info.presentationTimeUs, MediaCodec.BUFFER_FLAG_END_OF_STREAM)
                    }
                }
            }
        }
    }

    internal fun run(compressInfo: CompressInfo, containerParam: ContainerConverter.Param): Long {
        val deviceOption = deviceOption
        val adjustSize = getAdjustSize(compressInfo, deviceOption)

        containerParam.selectVideoIndex()
        containerParam.seekIfNeed(compressInfo.startTime)

        try {
            initCoders(containerParam, compressInfo, deviceOption)

            Observable.create(ObservableOnSubscribe<Any> {
                while (!inputDone) {
                    decodeFromExtractor(containerParam)
                }
            })
                    .subscribeOn(Schedulers.newThread())
                    .doOnError { throwable ->
                        if (throwable is Exception) {
                            throw throwable
                        }
                    }
                    .subscribe()

            while (!outputDone) {
                // write encoder data to muxer first, then read to encoder from decoder
                if (encodeToMuxer(containerParam, compressInfo)) {
                    if (!decoderDone) {
                        decoderToEncoder(compressInfo, deviceOption, adjustSize)
                    }
                }
            }
        } catch (e: Exception) {
            LogUtils.e(e)
            compressInfo.error = true
        }

        containerParam.unselectVideoIndex()
        release()
        return videoStartTime
    }

    private fun release() {
        if (outputSurface != null) {
            outputSurface!!.release()
        }
        if (inputSurface != null) {
            inputSurface!!.release()
        }
        if (decoder != null) {
            decoder!!.stop()
            decoder!!.release()
        }
        if (encoder != null) {
            encoder!!.stop()
            encoder!!.release()
        }
    }

    private class DeviceOption {
        internal var colorFormat: Int = 0
        internal var processorType = PROCESSOR_TYPE_OTHER
        internal var manufacturer = Build.MANUFACTURER.toLowerCase()
        internal var swapUV = 0
    }

    private class AdjustSize {
        internal var padding = 0
        internal var bufferSize: Int = 0
    }

    private fun getAdjustSize(compressInfo: CompressInfo, deviceOption: DeviceOption): AdjustSize {
        val adjustSize = AdjustSize()
        var resultHeightAligned = compressInfo.resultHeight
        adjustSize.bufferSize = compressInfo.resultWidth * compressInfo.resultHeight * 3 / 2

        if (deviceOption.processorType == PROCESSOR_TYPE_OTHER) {
            if (compressInfo.resultHeight % 16 != 0) {
                resultHeightAligned += 16 - compressInfo.resultHeight % 16
                adjustSize.padding = compressInfo.resultWidth * (resultHeightAligned - compressInfo.resultHeight)
                adjustSize.bufferSize += adjustSize.padding * 5 / 4
            }
        } else if (deviceOption.processorType == PROCESSOR_TYPE_QCOM) {
            if (deviceOption.manufacturer.toLowerCase() != "lge") {
                val uvoffset = compressInfo.resultWidth * compressInfo.resultHeight + 2047 and 2047.inv()
                adjustSize.padding = uvoffset - compressInfo.resultWidth * compressInfo.resultHeight
                adjustSize.bufferSize += adjustSize.padding
            }
        } else if (deviceOption.processorType == PROCESSOR_TYPE_TI) {
            //resultHeightAligned = 368;
            //bufferSize = resultWidth * resultHeightAligned * 3 / 2;
            //resultHeightAligned += (16 - (resultHeight % 16));
            //padding = resultWidth * (resultHeightAligned - resultHeight);
            //bufferSize += padding * 5 / 4;
        } else if (deviceOption.processorType == PROCESSOR_TYPE_MTK) {
            if (deviceOption.manufacturer == "baidu") {
                resultHeightAligned += 16 - compressInfo.resultHeight % 16
                adjustSize.padding = compressInfo.resultWidth * (resultHeightAligned - compressInfo.resultHeight)
                adjustSize.bufferSize += adjustSize.padding * 5 / 4
            }
        }

        return adjustSize
    }

    companion object {

        private val PROCESSOR_TYPE_OTHER = 0
        private val PROCESSOR_TYPE_QCOM = 1
        private val PROCESSOR_TYPE_INTEL = 2
        private val PROCESSOR_TYPE_MTK = 3
        private val PROCESSOR_TYPE_SEC = 4
        private val PROCESSOR_TYPE_TI = 5

        private val INVALID_VIDEO_TRACK_INDEX = -5
    }
}